lib/url.js

```diff
@@ -432,51 +282,35 @@ function urlResolveObject(source, relative) {
 }
 
 Url.prototype.resolveObject = function(relative) {
-  if (typeof relative === 'string') {
-    var rel = new Url();
-    rel.parse(relative, false, true);
-    relative = rel;
-  }
+  if (typeof relative === 'string')
+    relative = urlParse(relative, false, true);
 
-  var result = new Url();
-  var tkeys = Object.keys(this);
-  for (var tk = 0; tk < tkeys.length; tk++) {
-    var tkey = tkeys[tk];
-    result[tkey] = this[tkey];
-  }
+  var result = this._clone();
 
-  // hash is always overridden, no matter what.
-  // even href="" will remove it.
+  // Hash is always overridden, no matter what.
+  // even href='' will remove it.
   result.hash = relative.hash;
 
-  // if the relative url is empty, then there's nothing left to do here.
-  if (relative.href === '') {
-    result.href = result.format();
+  // If the relative url is empty, then there's nothing left to do here.
+  if (!relative.href) {
+    result._href = '';
     return result;
   }
 
-  // hrefs like //foo/bar always cut to the protocol.
-  if (relative.slashes && !relative.protocol) {
-    // take everything except the protocol from relative
-    var rkeys = Object.keys(relative);
-    for (var rk = 0; rk < rkeys.length; rk++) {
-      var rkey = rkeys[rk];
-      if (rkey !== 'protocol')
-        result[rkey] = relative[rkey];
-    }
+  // Hrefs like //foo/bar always cut to the protocol.
+  if (relative.slashes && !relative._protocol) {
+    relative._copyPropsTo(result, true);
 
-    //urlParse appends trailing / to urls like http://www.example.com
-    if (slashedProtocol[result.protocol] &&
+    if (_slashProtocols[result._protocol] &&
         result.hostname && !result.pathname) {
-      result.path = result.pathname = '/';
+      result.pathname = '/';
```

Comment:
I've got it!! my concern is this change breaks backward compatible.
But this pull request is tagged as semver-major. I didn't see this PR tag...
this change is reasonable. no problem.